### PR TITLE
Refactor: add learner into membership

### PR DESCRIPTION
**Pull request by lichuang**
Refactor: add learner in membership.

issue: https://github.com/datafuselabs/openraft/issues/44

1. add `learners` field in `Membership` struct,change `MessageSummary` impl to add learners info.
2. when call `add_learner`,append a membership change log.
3. after leader transfer, add nodes in `Membership::learners` into nodes,spawn replication stream for all learner nodes.
4. add `check_learner_after_leader_transfered` unit test to test when old leader has been removed,after new leader setup,the learner can receive logs in new cluster membership.
5. fix all the failed tests caused by `add_learner` append a log.
6. add tests in `Membership`.

### CODE REVIEW

File: openraft/src/core/admin.rs [hunk 1]
```diff
@@ -73,15 +73,48 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
...
+            return;
+        }
+        curr.add_learner(target);
+
+        // append_membership_log after learner has added into nodes
+        let new_config = self.core.effective_membership.membership.clone();
+
+        tracing::debug!(?new_config, "new_config");
+
+        let _ = self.append_membership_log(new_config, None).await;
```
Comment (drmingdrmer) [inline openraft/src/core/admin.rs:104]: `self.core.effective_membership` should only be updated by `append_payload_to_log()` which is called by `append_membership_log()`.

Before the log entry is written `effective_membership` should not be updated.

File: openraft/src/core/admin.rs [hunk 2]
```diff
@@ -73,15 +73,48 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
...
+            return;
+        }
+
+        let curr = &mut self.core.effective_membership.membership;
+
+        if curr.is_learner(target) {
+            tracing::debug!("target node {} is already a learner", target);
+
+            return;
+        }
```
Comment (drmingdrmer) [inline openraft/src/core/admin.rs:96]: Can these two condition checks be merged into one, such as `mermbers.contains(target)`?

File: openraft/src/membership/membership.rs [hunk 3]
```diff
@@ -89,6 +142,10 @@ impl Membership {
         false
     }
 
+    pub fn is_learner(&self, x: &NodeId) -> bool {
+        self.learners.contains(x)
+    }
```
Comment (drmingdrmer) [inline openraft/src/membership/membership.rs:147]: I believe these new APIs deserve a seat in `membershp_test.rs`.

File: openraft/src/membership/membership.rs [hunk 4]
```diff
@@ -44,12 +56,53 @@ impl Membership {
...
+
+    pub fn remove_learner(&mut self, id: &NodeId) {
+        self.learners.remove(id);
+    }
+
+    pub fn all_learners(&self) -> &BTreeSet<NodeId> {
+        &self.learners
     }
 
     pub fn all_nodes(&self) -> &BTreeSet<NodeId> {
```
Comment (drmingdrmer) [inline openraft/src/membership/membership.rs:108]: `all_nodes()` still returns only members. It should be renamed to `all_members()` or else.

**Review by drmingdrmer — COMMENTED** [review]
Not yet finished reading all of them.

The methods in `Membership` need to be refined to reflect its purpose correctly, and the corresponding test should be added or adjusted.

Another issue is that `change_membership()` should remove the node ids from `Membership.learners` if it is in the user specified node list.

And there should be a rule what to do when a `member` is removed from `Membership.members()`: The options are: turn it to a `learner`, or just remove it.

**Comment by lichuang** [comment]
> And there should be well-defined what to do when a member is removed from Membership.members(): The options are: turn it to a learner, or just remove it.

cool,maybe we can open an new issue to refactor this function.

File: openraft/src/membership/membership.rs [hunk 5]
```diff
@@ -19,51 +19,113 @@ use crate::NodeId;
...
+    }
+
+    pub fn new_single_with_learners(members: BTreeSet<NodeId>, learners: BTreeSet<NodeId>) -> Self {
+        let configs = vec![members];
+        let all_members = Self::build_all_members(&configs);
+        Membership {
+            learners,
+            configs,
+            all_members,
+        }
```
Comment (drmingdrmer) [inline openraft/src/membership/membership.rs:74]: There should be tests to ensure that  `replace()`, `push()`, `new_single_with_learners()` and `new_multi_with_learners()` won't store a `node_id` in both `configs` and `learners`

File: openraft/src/core/admin.rs [hunk 6]
```diff
@@ -73,15 +73,41 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
...
+        );
+
+        let curr = &self.core.effective_membership.membership;
+        if curr.contains(target) {
+            tracing::debug!(
+                "target node {} is already a member or learner,cannot add as learner",
+                target
+            );
+            return;
+        }
```
Comment (drmingdrmer) [inline openraft/src/core/admin.rs:91]: The result matters: If a learner already exists in the current config, it should notify the caller not to spawn a replication task for it.

File: openraft/src/core/admin.rs [hunk 7]
```diff
@@ -108,6 +136,33 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
...
+                Some(_node) => {
+                    continue;
+                }
+                None => {
+                    // if new member not include in the learners, there is pending membership change
+                    return true;
+                }
+            }
+        }
+        false
```
Comment (drmingdrmer) [inline openraft/src/core/admin.rs:165]: Why is it a `pending` state when there is a new member that is not in `learners`?

Comment (drmingdrmer) [reply openraft/src/core/admin.rs:165]: The first condition is to ensure no new membership would be proposed when there is uncommitted membership.
I do not quite understand why the second part is there. 🤔

Comment (lichuang) [reply openraft/src/core/admin.rs:171]: > The first condition is to ensure no new membership would be proposed when there is uncommitted membership. I do not quite understand why the second part is there. 🤔

in `Raft::change_membership`,first call `add_learner` and then `change_membership`,so when `add_learner` add a new log, `change_membership` cannot process in previous condition.

so i refactor the `change_member` condition judge into two case in function `has_pending_membership_change `:

1. if or not has a biggger commit log.
2. if there is new member not in the learner set.

File: openraft/src/membership/membership_test.rs [hunk 8]
```diff
@@ -58,9 +60,20 @@ fn test_membership_with_learners() -> anyhow::Result<()> {
...
         assert!(m1_23.is_learner(&3));
+
+        // test replace and push, does not affect learners
+        m1_2.replace(vec![btreeset! {3,4}]);
+        assert_eq!(&btreeset! {3,4}, m1_2.all_members());
+        assert_eq!(&btreeset! {2}, m1_2.all_learners());
+
+        m1_2.push(btreeset! {5});
+        assert_eq!(&btreeset! {3,4,5}, m1_2.all_members());
+        assert_eq!(&btreeset! {2}, m1_2.all_learners());
```
Comment (drmingdrmer) [inline openraft/src/membership/membership_test.rs:71]: replace/push test should be put in `test_membership_update()`.

**Review by drmingdrmer — COMMENTED** [review]
Is there a test for  `new_multi_with_learners()` with overlapping `members` and `learners`?

File: openraft/tests/client_api/main.rs [hunk 9]
```diff
@@ -5,6 +5,6 @@ mod fixtures;
 // The number indicate the preferred running order for these case.
 // The later tests may depend on the earlier ones.
 
-mod t10_client_writes;
-mod t20_client_reads;
+//mod t10_client_writes;
+//mod t20_client_reads;
```
[the remark this answers is not in the document]
Comment (lichuang) [reply openraft/tests/client_api/main.rs]: > Why removed these test?

commit my debug changes...

File: openraft/src/core/admin.rs (hunk 7 above)
Comment (drmingdrmer) [reply openraft/src/core/admin.rs:171]: It seems we should add learners and change members in one log entry.
Then there won't be such a problem.

Another issue is that we need to spawn replication tasks just after `append_payload_to_log()` and before `replicate_client_request()`.

What say you man?

File: openraft/tests/client_api/main.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply openraft/tests/client_api/main.rs]: With Clion it is able to run a single test function or a single test file. Very convenient!!
You deserve it!

File: openraft/src/core/admin.rs (hunk 7 above)
Comment (lichuang) [reply openraft/src/core/admin.rs:171]: > It seems we should add learners and change members in one log entry. Then there won't be such a problem.

yes,i agree with this,i will refactor the source.

> Another issue is that we need to spawn replication tasks just after append_payload_to_log() and before replicate_client_request().

ok, assign a new issue to me.

**Review by drmingdrmer — APPROVED** [review]

**Pull request merged**